wgsl/extract-grammar.py

```diff
@@ -433,236 +368,557 @@ def grammar_from_rule(key, value):
+                YEAR=date.today().year) + grammar_source + "\n")
+            grammar_file.close()
+
+    if input_bs_is_fresh:
+        # Save scanned lines for next time.
+        with open(previously_scanned_bs_file,"w") as previous_file:
+            for line in scan_result['raw']:
+                previous_file.write(line)
+
+    with open(os.path.join(options.grammar_dir,"package.json"), "w") as grammar_package:
+        grammar_package.write('{\n')
+        grammar_package.write('    "name": "tree-sitter-wgsl",\n')
+        grammar_package.write('    "dependencies": {\n')
+        grammar_package.write('        "nan": "^2.15.0"\n')
+        grammar_package.write('    },\n')
+        grammar_package.write('    "devDependencies": {\n')
+        grammar_package.write('        "tree-sitter-cli": "^0.20.7"\n')
+        grammar_package.write('    },\n')
+        grammar_package.write('    "main": "bindings/node"\n')
+        grammar_package.write('}\n')
+
+    return True
+
+def flow_build(options):
+    """
+    Build the shared library for the custom tree-sitter scanner.
+    """
+
+    print("{}: Build...".format(options.script))
+    if not os.path.exists(options.grammar_filename):
+        print("missing grammar file: {}")
+        return False
+
+    # External scanner for nested block comments
+    # For the API, see https://tree-sitter.github.io/tree-sitter/creating-parsers#external-scanners
+    # See: https://github.com/tree-sitter/tree-sitter-rust/blob/master/src/scanner.c
+
+    os.makedirs(os.path.join(options.grammar_dir, "src"), exist_ok=True)
+
+    # Remove the old custom scanner, if it exists.
+    scanner_c_staging = os.path.join(options.grammar_dir, "src", "scanner.c")
+    if os.path.exists(scanner_c_staging):
+        os.remove(scanner_c_staging)
+    # Copy the new scanner into place, if newer
+    scanner_cc_staging = os.path.join(options.grammar_dir, "src", "scanner.cc")
+    if newer_than(options.scanner_cc_filename, scanner_cc_staging):
+        shutil.copyfile(options.scanner_cc_filename, scanner_cc_staging)
+
+
+    # Use "npm install" to create the tree-sitter CLI that has WGSL
+    # support.  But "npm install" fetches data over the network.
+    # That can be flaky, so only invoke it when needed.
+    if os.path.exists("grammar/node_modules/tree-sitter-cli") and os.path.exists("grammar/node_modules/nan"):
+        # "npm install" has been run already.
+        pass
+    else:
+        subprocess.run(["npm", "install"], cwd=options.grammar_dir, check=True)
+    subprocess.run(["npx", "tree-sitter", "generate"],
+                   cwd=options.grammar_dir, check=True)
+    # Following are commented for future reference to expose playground
+    # Remove "--docker" if local environment matches with the container
+    # subprocess.run(["npx", "tree-sitter", "build-wasm", "--docker"],
+    #                cwd=options.grammar_dir, check=True)
+
+
+    def build_library(output_file, input_files):
```

Comment:
py-tree-sitter does not know how to compile C++ files with Apple Clang that I have.  Clang++ doesn't default to C++17, but I need to supply those arguments.  So invoke the compiler ourselves with the proper flags.

There is an upstream PR that's 2 years old to do essentially this.  I can't wait for that.